[Experimental] Support AMD GPU via HIP (ROCm2.7.0+) (#1094)
cupy/cupy: docs/source/install_rocm.rst

* AMD GPU
* `ROCm <https://rocm.github.io/install.html>`_
    * Supported Versions: ROCm 2.6+.
* `Python <https://python.org/>`_
* `NumPy <http://www.numpy.org/>`_

And please install ROCm libraries.

::

  $ sudo apt install hipblas hipsparse rocrand

Before installing CuPy, we recommend you to upgrade ``setuptools`` and ``pip``::

  $ pip install -U setuptools pip

Install CuPy from Source
------------------------

It is recommended to use wheels whenever possible.
However, if wheels cannot meet your requirements (e.g., you are running non-Linux environment or want to use a version of CUDA / cuDNN / NCCL not supported by wheels), you can also build CuPy from source.

When installing from source, C++ compiler such as ``g++`` is required.
You need to install it before installing CuPy.
This is typical installation method for each platform::

  # Ubuntu 16.04
  $ apt-get install g++

.. note::

   If you upgrade or downgrade the version of ROCm, you may need to reinstall CuPy.
   See :ref:`rocm_install_reinstall` for details.

Using pip
~~~~~~~~~

You can install `CuPy package <https://pypi.python.org/pypi/cupy>`_ via ``pip``.

::

>   $ export HCC_AMDGPU_TARGET=gfx900

Reviewer: ```suggestion
  $ export HCC_AMDGPU_TARGET=gfx900  # This value should be changed based on your GPU
```

  $ export __HIP_PLATFORM_HCC__
  $ export CUPY_INSTALL_USE_HIP=1
  $ pip install cupy

Using Tarball
~~~~~~~~~~~~~

The tarball of the source tree is available via ``pip download cupy`` or from `the release notes page <https://github.com/cupy/cupy/releases>`_.
You can install CuPy from the tarball::

  $ pip install cupy-x.x.x.tar.gz

You can also install the development version of CuPy from a cloned Git repository::

  $ git clone https://github.com/cupy/cupy.git
  $ cd cupy

>   $ export HCC_AMDGPU_TARGET=gfx900

Reviewer: The value of `HCC_AMDGPU_TARGET`, in this case, `gfx900` will be different for different AMD GPUs, right? If so, I think it's better to mention users have to change it according to the GPU.
```suggestion
  $ export HCC_AMDGPU_TARGET=gfx900  # This value should be changed based on your GPU
```

  $ export __HIP_PLATFORM_HCC__
  $ export CUPY_INSTALL_USE_HIP=1
  $ pip install .

> If you are using source tree downloaded from GitHub, you need to install Cython 0.28.0 or later (``pip install cython``).

Reviewer: ```suggestion
If you are using the source tree downloaded from GitHub, you need to install Cython 0.28.0 or later (``pip install cython``).
```

Uninstall CuPy
--------------

Use pip to uninstall CuPy::

  $ pip uninstall cupy

.. note::

   When you upgrade Chainer, ``pip`` sometimes installs the new version without removing the old one in ``site-packages``.
   In this case, ``pip uninstall`` only removes the latest one.
   To ensure that CuPy is completely removed, run the above command repeatedly until ``pip`` returns an error.

Upgrade CuPy
------------

Just use ``pip install`` with ``-U`` option::

>   $ export HCC_AMDGPU_TARGET=gfx900

Reviewer: ```suggestion
  $ export HCC_AMDGPU_TARGET=gfx900  # This value should be changed based on your GPU
```

  $ export __HIP_PLATFORM_HCC__
  $ export CUPY_INSTALL_USE_HIP=1
  $ pip install -U cupy

.. _rocm_install_reinstall:

Reinstall CuPy
--------------

> If you want to reinstall CuPy, please uninstall CuPy and then install it.

Reviewer: ```suggestion
If you want to reinstall CuPy, please uninstall CuPy first, and then install again.
```

When reinstalling CuPy, we recommend to use ``--no-cache-dir`` option as ``pip`` caches the previously built binaries::

  $ pip uninstall cupy

>   $ export HCC_AMDGPU_TARGET=gfx900

Reviewer: ```suggestion
  $ export HCC_AMDGPU_TARGET=gfx900  # This value should be changed based on your GPU
```

  $ export __HIP_PLATFORM_HCC__
  $ export CUPY_INSTALL_USE_HIP=1
  $ pip install cupy --no-cache-dir

FAQ
---

.. _rocm_install_error:

``pip`` fails to install CuPy
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Please make sure that you are using the latest ``setuptools`` and ``pip``::

  $ pip install -U setuptools pip

> Use ``-vvvv`` option with ``pip`` command.

Reviewer: ```suggestion
Use ``-vvvv`` option with ``pip`` command to investigate the details of errors.
```
